File: java-util/src/main/java/eu/janvdb/aocutil/java/Point3D.java
package eu.janvdb.aocutil.java;

import java.util.Objects;

public class Point3D {

	private final int x, y, z;

	public Point3D(int x, int y, int z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}

	public int getX() {
		return x;
	}

	public int getY() {
		return y;
	}

	public int getZ() {
		return z;
	}

	public int getManhattanDistanceFromOrigin() {
		return Math.abs(x) + Math.abs(y) + Math.abs(z);
	}

	public int getManhattanDistance(Point3D other) {
		return Math.abs(x - other.x) + Math.abs(y - other.y) + Math.abs(z - other.z);
	}

	@Override
	public String toString() {
		return "(" + x + ", " + y + ", " + z + ')';
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Point3D other = (Point3D) o;
		return x == other.x && y == other.y && z == other.z;
	}

	@Override
	public int hashCode() {
		return Objects.hash(x, y, z);
	}

	public Point3D add(Point3D other) {
		return new Point3D(x + other.x, y + other.y, z + other.z);
	}

	public Point3D add(int x1, int y1, int z1) {
		return new Point3D(x + x1, y + y1, z + z1);
	}

	public Point3D minus(Point3D other) {
		return new Point3D(x - other.x, y - other.y, z - other.z);
	}

	public Point3D move(int dx, int dy, int dz) {
		return new Point3D(x+dx, y+dy, z+dz);
	}
}

File: kotlin-util/src/main/kotlin/eu/janvdb/aocutil/kotlin/point3d/Point3D.kt
package eu.janvdb.aocutil.kotlin.point3d

import kotlin.math.abs

data class Point3D(val x: Int, val y: Int, val z: Int) {
	operator fun plus(other: Point3D): Point3D = Point3D(x + other.x, y + other.y, z + other.z)
	operator fun minus(other: Point3D) = Point3D(x - other.x, y - other.y, z - other.z)

	fun manhattanDistance(other: Point3D) = abs(x - other.x) + abs(y - other.y) + abs(z - other.z)

    fun distanceSquared(other: Point3D) =
        1L * (x - other.x) * (x - other.x) + 1L * (y - other.y) * (y - other.y) + 1L * (z - other.z) * (z - other.z)

	fun move(dx: Int, dy: Int, dz: Int): Point3D {
		return Point3D(x + dx, y + dy, z + dz)
	}

	companion object {
		fun createCommaSeparated(s: String): Point3D {
			val pair = s.split(",")
			return Point3D(pair[0].trim().toInt(), pair[1].trim().toInt(), pair[2].trim().toInt())
		}
	}
}
